File: android/src/main/kotlin/com/gdelataillade/alarm/models/AlarmSettings.kt
package com.gdelataillade.alarm.models

import com.gdelataillade.alarm.generated.AlarmSettingsWire
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.*
import java.time.Duration
import kotlin.time.toKotlinDuration
import java.util.Date

@Serializable
data class AlarmSettings(
    val id: Int,
    @Serializable(with = DateSerializer::class)
    val dateTime: Date,
    val assetAudioPath: String,
    val volumeSettings: VolumeSettings,
    val notificationSettings: NotificationSettings,
    val loopAudio: Boolean,
    val vibrate: Boolean,
    val warningNotificationOnKill: Boolean,
    val androidFullScreenIntent: Boolean,
    val allowAlarmOverlap: Boolean = false, // Defaults to false for backward compatibility
    val androidStopAlarmOnTermination: Boolean = true, // Defaults to true for backward compatibility
) {
    companion object {
        fun fromWire(e: AlarmSettingsWire): AlarmSettings {
            return AlarmSettings(
                e.id.toInt(),
                Date(e.millisecondsSinceEpoch),
                e.assetAudioPath,
                VolumeSettings.fromWire(e.volumeSettings),
                NotificationSettings.fromWire(e.notificationSettings),
                e.loopAudio,
                e.vibrate,
                e.warningNotificationOnKill,
                e.androidFullScreenIntent,
                e.allowAlarmOverlap,
                e.androidStopAlarmOnTermination,
            )
        }

        /**
         * Handles backward compatibility for missing fields like `volumeSettings` and `allowAlarmOverlap`.
         */
        fun fromJson(json: String): AlarmSettings {
            val jsonObject = Json.parseToJsonElement(json).jsonObject

            val id = jsonObject.primitiveInt("id") ?: throw SerializationException("Missing 'id'")
            val dateTimeMillis = jsonObject.primitiveLong("dateTime") ?: throw SerializationException("Missing 'dateTime'")
            val assetAudioPath = jsonObject.primitiveString("assetAudioPath") ?: throw SerializationException("Missing 'assetAudioPath'")
            val notificationSettings = jsonObject["notificationSettings"]?.let {
                Json.decodeFromJsonElement(NotificationSettings.serializer(), it)
            } ?: throw SerializationException("Missing 'notificationSettings'")
            val loopAudio = jsonObject.primitiveBoolean("loopAudio") ?: throw SerializationException("Missing 'loopAudio'")
            val vibrate = jsonObject.primitiveBoolean("vibrate") ?: throw SerializationException("Missing 'vibrate'")
            val warningNotificationOnKill = jsonObject.primitiveBoolean("warningNotificationOnKill")
                ?: throw SerializationException("Missing 'warningNotificationOnKill'")
            val androidFullScreenIntent = jsonObject.primitiveBoolean("androidFullScreenIntent")
                ?: throw SerializationException("Missing 'androidFullScreenIntent'")

            // Handle backward compatibility for `allowAlarmOverlap`
            val allowAlarmOverlap = jsonObject.primitiveBoolean("allowAlarmOverlap") ?: false

            // Handle backward compatibility for `androidStopAlarmOnTermination`
            val androidStopAlarmOnTermination = jsonObject.primitiveBoolean("androidStopAlarmOnTermination") ?: true

            // Handle backward compatibility for `volumeSettings`
            val volumeSettings = jsonObject["volumeSettings"]?.let {
                Json.decodeFromJsonElement(VolumeSettings.serializer(), it)
            } ?: run {
                val volume = jsonObject.primitiveDouble("volume")
                val fadeDurationSeconds = jsonObject.primitiveDouble("fadeDuration")
                val fadeDuration = fadeDurationSeconds?.let { Duration.ofMillis((it * 1000).toLong()) }
                val volumeEnforced = jsonObject.primitiveBoolean("volumeEnforced") ?: false

                VolumeSettings(
                    volume = volume,
                    fadeDuration = fadeDuration?.toKotlinDuration(),
                    fadeSteps = emptyList(), // No equivalent for older models
                    volumeEnforced = volumeEnforced
                )
            }

            return AlarmSettings(
                id = id,
                dateTime = Date(dateTimeMillis),
                assetAudioPath = assetAudioPath,
                volumeSettings = volumeSettings,
                notificationSettings = notificationSettings,
                loopAudio = loopAudio,
                vibrate = vibrate,
                warningNotificationOnKill = warningNotificationOnKill,
                androidFullScreenIntent = androidFullScreenIntent,
                allowAlarmOverlap = allowAlarmOverlap,
                androidStopAlarmOnTermination = androidStopAlarmOnTermination,
            )
        }
    }
}

/**
 * Custom serializer for Java's `Date` type.
 */
object DateSerializer : KSerializer<Date> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Date", PrimitiveKind.LONG)
    override fun serialize(encoder: Encoder, value: Date) = encoder.encodeLong(value.time)
    override fun deserialize(decoder: Decoder): Date = Date(decoder.decodeLong())
}

// Extension functions for safer primitive extraction from JsonObject
private fun JsonObject.primitiveInt(key: String): Int? = this[key]?.jsonPrimitive?.content?.toIntOrNull()
private fun JsonObject.primitiveLong(key: String): Long? = this[key]?.jsonPrimitive?.content?.toLongOrNull()
private fun JsonObject.primitiveDouble(key: String): Double? = this[key]?.jsonPrimitive?.content?.toDoubleOrNull()
private fun JsonObject.primitiveString(key: String): String? = this[key]?.jsonPrimitive?.content
private fun JsonObject.primitiveBoolean(key: String): Boolean? = this[key]?.jsonPrimitive?.content?.toBooleanStrictOrNull()

File: ios/Classes/models/AlarmSettings.swift
import Foundation

struct AlarmSettings: Codable {
    let id: Int
    let dateTime: Date
    let assetAudioPath: String
    let volumeSettings: VolumeSettings
    let notificationSettings: NotificationSettings
    let loopAudio: Bool
    let vibrate: Bool
    let warningNotificationOnKill: Bool
    let androidFullScreenIntent: Bool
    let allowAlarmOverlap: Bool
    let iOSBackgroundAudio: Bool

    enum CodingKeys: String, CodingKey {
        case id, dateTime, assetAudioPath, volumeSettings, notificationSettings,
             loopAudio, vibrate, warningNotificationOnKill, androidFullScreenIntent,
             allowAlarmOverlap, iOSBackgroundAudio, volume, fadeDuration, volumeEnforced
    }

    /// Custom initializer to handle backward compatibility for older models
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        // Decode mandatory fields
        id = try container.decode(Int.self, forKey: .id)
        dateTime = try container.decode(Date.self, forKey: .dateTime)
        assetAudioPath = try container.decode(String.self, forKey: .assetAudioPath)
        notificationSettings = try container.decode(NotificationSettings.self, forKey: .notificationSettings)
        loopAudio = try container.decode(Bool.self, forKey: .loopAudio)
        vibrate = try container.decode(Bool.self, forKey: .vibrate)
        warningNotificationOnKill = try container.decode(Bool.self, forKey: .warningNotificationOnKill)
        androidFullScreenIntent = try container.decode(Bool.self, forKey: .androidFullScreenIntent)

        // Backward compatibility for `allowAlarmOverlap`
        allowAlarmOverlap = try container.decodeIfPresent(Bool.self, forKey: .allowAlarmOverlap) ?? false
        
        // Backward compatibility for `iOSBackgroundAudio`
        iOSBackgroundAudio = try container.decodeIfPresent(Bool.self, forKey: .iOSBackgroundAudio) ?? true

        // Backward compatibility for `volumeSettings`
        if let volumeSettingsDecoded = try? container.decode(VolumeSettings.self, forKey: .volumeSettings) {
            volumeSettings = volumeSettingsDecoded
        } else {
            // Reconstruct `volumeSettings` from older fields
            let volume = try container.decodeIfPresent(Double.self, forKey: .volume)
            let fadeDurationSeconds = try container.decodeIfPresent(Double.self, forKey: .fadeDuration)
            let fadeDuration = fadeDurationSeconds.map { TimeInterval($0) }
            let volumeEnforced = try container.decodeIfPresent(Bool.self, forKey: .volumeEnforced) ?? false

            volumeSettings = VolumeSettings(
                volume: volume,
                fadeDuration: fadeDuration,
                fadeSteps: [], // No equivalent for fadeSteps in older models
                volumeEnforced: volumeEnforced
            )
        }
    }

    /// Encode method to support `Encodable` protocol
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        try container.encode(id, forKey: .id)
        try container.encode(dateTime, forKey: .dateTime)
        try container.encode(assetAudioPath, forKey: .assetAudioPath)
        try container.encode(volumeSettings, forKey: .volumeSettings)
        try container.encode(notificationSettings, forKey: .notificationSettings)
        try container.encode(loopAudio, forKey: .loopAudio)
        try container.encode(vibrate, forKey: .vibrate)
        try container.encode(warningNotificationOnKill, forKey: .warningNotificationOnKill)
        try container.encode(androidFullScreenIntent, forKey: .androidFullScreenIntent)
        try container.encode(allowAlarmOverlap, forKey: .allowAlarmOverlap)
        try container.encode(iOSBackgroundAudio, forKey: .iOSBackgroundAudio)
    }

    /// Memberwise initializer
    init(
        id: Int,
        dateTime: Date,
        assetAudioPath: String,
        volumeSettings: VolumeSettings,
        notificationSettings: NotificationSettings,
        loopAudio: Bool,
        vibrate: Bool,
        warningNotificationOnKill: Bool,
        androidFullScreenIntent: Bool,
        allowAlarmOverlap: Bool,
        iOSBackgroundAudio: Bool
    ) {
        self.id = id
        self.dateTime = dateTime
        self.assetAudioPath = assetAudioPath
        self.volumeSettings = volumeSettings
        self.notificationSettings = notificationSettings
        self.loopAudio = loopAudio
        self.vibrate = vibrate
        self.warningNotificationOnKill = warningNotificationOnKill
        self.androidFullScreenIntent = androidFullScreenIntent
        self.allowAlarmOverlap = allowAlarmOverlap
        self.iOSBackgroundAudio = iOSBackgroundAudio
    }

    /// Converts from wire model to `AlarmSettings`.
    static func from(wire: AlarmSettingsWire) -> AlarmSettings {
        return AlarmSettings(
            id: Int(truncatingIfNeeded: wire.id),
            dateTime: Date(timeIntervalSince1970: TimeInterval(wire.millisecondsSinceEpoch / 1_000)),
            assetAudioPath: wire.assetAudioPath,
            volumeSettings: VolumeSettings.from(wire: wire.volumeSettings),
            notificationSettings: NotificationSettings.from(wire: wire.notificationSettings),
            loopAudio: wire.loopAudio,
            vibrate: wire.vibrate,
            warningNotificationOnKill: wire.warningNotificationOnKill,
            androidFullScreenIntent: wire.androidFullScreenIntent,
            allowAlarmOverlap: wire.allowAlarmOverlap,
            iOSBackgroundAudio: wire.iOSBackgroundAudio
        )
    }
}
